catalog/rocks/refresolver.go

```diff
@@ -0,0 +1,132 @@
+package rocks
+
+import (
+	"context"
+	"errors"
+
+	"github.com/treeverse/lakefs/ident"
+)
+
+type RefStore interface {
+	GetBranch(ctx context.Context, repositoryID RepositoryID, branchID BranchID) (*Branch, error)
+	GetCommit(ctx context.Context, repositoryID RepositoryID, commitID CommitID) (*Commit, error)
+	Log(ctx context.Context, repositoryID RepositoryID, from CommitID) (CommitIterator, error)
+}
+
+type reference struct {
+	typ      ReferenceType
+	branch   *Branch
+	commitID *CommitID
+}
+
+func (r reference) Type() ReferenceType {
+	return r.typ
+}
+
+func (r reference) Branch() Branch {
+	return *r.branch
+}
+
+func (r reference) CommitID() CommitID {
+	return *r.commitID
+}
+
+func ResolveRef(ctx context.Context, store RefStore, repositoryID RepositoryID, ref Ref) (Reference, error) {
+	// first we need to parse-rev to get a list references
+	// valid revs: branch, tag, commit ID, commit ID prefix (as long as unambiguous)
+	// valid modifiers: ~N
+	parsed, err := RevParse(ref)
+	if err != nil {
+		return nil, err
+	}
+
+	var baseCommit CommitID
+	if isAHash(parsed.BaseRev) {
+		commit, err := store.GetCommit(ctx, repositoryID, CommitID(parsed.BaseRev))
+		if err != nil && !errors.Is(err, ErrNotFound) {
+			// couldn't check if it's a commit
+			return nil, err
+		}
+		if err == nil {
+			baseCommit = CommitID(ident.ContentAddress(commit))
+		}
+		// otherwise, simply not a commit. Moving on.
+	}
+
+	if baseCommit == "" {
+		// check if it's a branch
+		branch, err := store.GetBranch(ctx, repositoryID, BranchID(parsed.BaseRev))
+		if err != nil && !errors.Is(err, ErrNotFound) {
+			return nil, err
+		}
+		if err == nil {
+			baseCommit = branch.CommitID
+		}
+
+		if err == nil && len(parsed.Modifiers) == 0 {
```

Comment:
Ya, it hides the private field - staging token.
Only a branch, and only at its tip will have that - this is why we return a branch object (that contains it, albeit privately) instead of resolving it to the underlying commit.